Mathlib/Combinatorics/Flow/Basic.lean

```diff
@@ -0,0 +1,232 @@
+  nonneg : ∀ u v, 0 ≤ cap u v
+  cap_zero_of_not_adj : ∀ u v, ¬Adj u v → cap u v = 0 := by intro u v; aesop
+
+/-- A `Network` where all capacities are symmetrical. -/
+structure UndirectedNetwork extends Network V R where
+  symm : ∀ u v, cap u v = cap v u
+
+end NetworkDefinition
+
+variable {V : Type u_V} {R : Type u_R}
+
+namespace Network
+
+section PseudoFlowDefinition
+
+variable [Zero R] [Preorder R]
+
+/-- A `PseudoFlow` is an assignment of non-negative flow values to each arc of the network, such
+that the assigned amount does not exceed the capacity.
+
+Note that this definition, in contrast to what can sometimes be found in the literature, does
+not require that $f(u, v) = -f(v, u)$. This extra requirement eliminates cycles between two
+vertices and makes some theorem statements shorter. However, since larger cycles are still
+permitted, this usually does not help in proofs. To the contrary, defining a flow becomes more
+cumbersome, sometimes significantly so. -/
+@[ext]
+structure PseudoFlow (N : Network V R) where
+  /-- The assignment of flow values to each arc.
+  Do NOT use. Use the coercion to function instead. -/
+  toFun : V → V → R
+  nonneg' : ∀ u v, 0 ≤ toFun u v
+  capacity' : ∀ u v, toFun u v ≤ N.cap u v
+
+namespace PseudoFlow
+
+variable {N : Network V R}
+
+instance instFunLike : FunLike N.PseudoFlow V (V → R) where
+  coe := PseudoFlow.toFun
+  coe_injective' f g h := by cases f; cases g; congr
+
+variable (f : N.PseudoFlow)
+
+lemma nonneg : ∀ u v, 0 ≤ f u v := f.nonneg'
+
+lemma capacity : ∀ u v, f u v ≤ N.cap u v := f.capacity'
+
+instance : Zero N.PseudoFlow where
+  zero.toFun _ _ := 0
+  zero.nonneg' _ _ := le_rfl
+  zero.capacity' := N.nonneg
+
+end PseudoFlow
+end PseudoFlowDefinition
+
+variable [Fintype V]
+
+namespace PseudoFlow
+
+variable [Preorder R] [AddCommMonoid R] {N : Network V R} (f : N.PseudoFlow) (v : V)
+
+/-- The incoming flow of a vertex. --/
+def incoming := ∑ u, f u v
+/-- The outgoing flow of a vertex. --/
+def outgoing := ∑ w, f v w
+/-- The excess flow of a vertex (the incoming flow minus the outgoing flow). --/
+def excess [Sub R] := f.incoming v - f.outgoing v
+
+end PseudoFlow
+
+section PreFlowDefinition
+
+variable [Preorder R] [AddCommMonoid R] (N : Network V R) (s t : V)
+
+/-- A `PreFlow` from `s` to `t` is a `PseudoFlow` such that all vertices except `s` and `t` have
+non-negative excess, that is, the outgoing flow is at most as much as the incoming flow. -/
+@[ext]
+structure PreFlow extends N.PseudoFlow where
+  excess_nonneg : ∀ v, v ≠ s ∧ v ≠ t → toPseudoFlow.outgoing v ≤ toPseudoFlow.incoming v
+
+namespace PreFlow
+
+variable {N s t}
+
+instance : Zero (N.PreFlow s t) where
+  zero := { (0 : N.PseudoFlow) with excess_nonneg := fun _ _ ↦ le_rfl }
+
+/-- The value of a flow is the amount of flow units that leave the source vertex and arrive at the
+sink vertex. In particular, this is equal to the excess of `t`. -/
+def value [Sub R] (f : N.PreFlow s t) := f.excess t
```

Comment:
It looks to me like this definition will also work for a pseudoflow, without changes? Perhaps it's less meaningful there, but it seems perfectly well-defined.

Reply:
I thought about this too, but since pseudoflows don't have any associated source or sink in the current definition, there is no vertex to fix for the definition of `value`. This of course raises the question whether `PseudoFlow` should have these vertices, but I figured that if it is not needed in the definition, it doesn't really make sense (I don't think that there will be much theory regarding `PseudoFlow` directly, so it will probably be easy to change the definition either way later)

One interesting observation I had when splitting the definitions for `Flow` into `PseudoFlow` and `PreFlow` is that for `Flow`, both `-f.excess s` and `f.excess t` make perfect sense for `value` (and I find the former more intuitive), but with preflows, only `f.excess t` makes sense. Without the restrictions given by `PreFlow`, there is no vertex left with a sensible excess to pick for `value`

What do you think?